### PR TITLE
remove RDC flags when using CMake language CUDA

### PR DESCRIPTION
**Pull request by ibaned**
When compiling with CUDA as a CMake language,
the RDC aspect is handled by CMake's
CUDA_SEPARABLE_COMPILATION property on targets and the associated CMAKE_CUDA_SEPARABLE_COMPILATION
variable that sets the default value for that
property.
The arguments set by this part of Kokkos' CMake
actually conflict with the above.

### CODE REVIEW

**Comment by ibaned** [comment]
> Should `KOKKOS_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE=ON` then set `CMAKE_CUDA_SEPARABLE_COMPILATION=ON` when compiling with `CUDA` as a `CMake` language?

Yes, that makes sense! Can you help me identify where in the Kokkos CMake files that should be set? I can add it to this PR then if you'd like.

**Comment by crtrott** [comment]
Yeah agreed, we should set that option then, and/or warn if there is a mismatch of that with the Kokkos setting.

**Comment by ibaned** [comment]
Okay, I added some code to ensure that, if we are building using CMake's CUDA language, then setting either of those options on will set both of them on. Let me know if anything else is missing from this.

File: cmake/kokkos_enable_options.cmake [hunk 1]
```diff
@@ -130,6 +130,24 @@ IF (KOKKOS_ENABLE_AGGRESSIVE_VECTORIZATION)
...
+# Force consistency of KOKKOS_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE
+# and CMAKE_CUDA_SEPARABLE_COMPILATION when we are compiling
+# using the CMake CUDA language support.
+# Either one being on will turn the other one on.
+IF (KOKKOS_COMPILE_LANGUAGE STREQUAL CUDA)
+  IF (KOKKOS_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE)
+    IF (CMAKE_CUDA_SEPARABLE_COMPILATION)
+    ELSE()
+      SET(CMAKE_CUDA_SEPARABLE_COMPILATION ON)
+    ENDIF()
```
Comment (masterleinad) [inline cmake/kokkos_enable_options.cmake:142]:
```suggestion
    IF (NOT CMAKE_CUDA_SEPARABLE_COMPILATION)
      SET(CMAKE_CUDA_SEPARABLE_COMPILATION ON)
    ENDIF()
```

File: cmake/kokkos_enable_options.cmake [hunk 2]
```diff
@@ -130,6 +130,24 @@ IF (KOKKOS_ENABLE_AGGRESSIVE_VECTORIZATION)
...
+  ELSE()
+    IF (CMAKE_CUDA_SEPARABLE_COMPILATION)
+      SET(KOKKOS_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE ON)
+    ELSE()
+    ENDIF()
```
Comment (masterleinad) [inline cmake/kokkos_enable_options.cmake:147]:
```suggestion
    IF (CMAKE_CUDA_SEPARABLE_COMPILATION)
      SET(KOKKOS_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE ON)
    ENDIF()
```

File: cmake/kokkos_arch.cmake [hunk 3]
```diff
@@ -450,11 +450,13 @@ IF (KOKKOS_ARCH_POWER9)
 ENDIF()
 
 IF (KOKKOS_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE)
-  COMPILER_SPECIFIC_FLAGS(
-    Clang  -fcuda-rdc
-    NVIDIA --relocatable-device-code=true
-    NVHPC -gpu=rdc
-  )
+  IF (NOT KOKKOS_COMPILE_LANGUAGE STREQUAL CUDA)
```
Comment (dalg24) [inline cmake/kokkos_arch.cmake:453]: This does not seem right
Assume RDC is OFF but we are using NVHPC L462 we'd still be adding the `-gpu=nordc` flag

File: cmake/kokkos_enable_options.cmake [hunk 4]
```diff
@@ -130,6 +130,22 @@ IF (KOKKOS_ENABLE_AGGRESSIVE_VECTORIZATION)
...
 
+# Force consistency of KOKKOS_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE
+# and CMAKE_CUDA_SEPARABLE_COMPILATION when we are compiling
+# using the CMake CUDA language support.
+# Either one being on will turn the other one on.
+IF (KOKKOS_COMPILE_LANGUAGE STREQUAL CUDA)
+  IF (KOKKOS_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE)
+    IF (NOT CMAKE_CUDA_SEPARABLE_COMPILATION)
+      SET(CMAKE_CUDA_SEPARABLE_COMPILATION ON)
+    ENDIF()
```
Comment (dalg24) [inline cmake/kokkos_enable_options.cmake:141]: I think we might want to raise a warning (or display a STATUS message) when setting the `CMAKE_CUDA_SEPARABLE_COMPILATION` variable

File: cmake/kokkos_arch.cmake (hunk 3 above)
Comment (masterleinad) [reply cmake/kokkos_arch.cmake:453]: I agree. This if condition should guard from setting any flags.

**Comment by ibaned** [comment]
Thank you @masterleinad and @dalg24 for your reviews! I believe I've just addressed both of your suggestions, please let me know if other changes are desired.

**Review by masterleinad — APPROVED** [review]

**Review by dalg24 — APPROVED** [review]

**Pull request merged**